# Context

### Novel Coronavirus 2019 (nCoV-2019) is a virus which affects respiratory system and was first discovered in wuhan, China. Some early reports suggested that virus may have been transmitted from animal to person. As we know whole world has been shutdown  because of the widespread cases. At this time it's unclear how easily or sustainably this virus is spreading between people.

### So to start our machine learning project we will use dataset we got from Kaggle.This dataset is being used because this dataset is local to Canada.

<a href = "https://www.kaggle.com/ryanxjhan/coronaviruscovid19-canada">Please Click here for the link to dataset.</a>

# Current Cases (WorldWide)
### To know how bad the world has been affected lets get some information on current situation.

#### Lets import all the dependencies for scrapping the website

In [1]:
import bs4
from urllib.request import Request, urlopen
from urllib.request import urlopen as uReq
from bs4 import BeautifulSoup as soup
import pandas as pd

In [2]:
# grabbing the url

url = "https://ncov2019.live/"
req = Request(url, headers={"User-Agent" : "Mozilla/5.0"})

webpage = urlopen(req).read()

#parsing it as lxml
pagesoup = soup(webpage,"lxml")

#### Website Information
1. Website Name
2. Link to Website

In [3]:
from IPython.display import display, Markdown

In [4]:
#finding the relevant tags to scrap the data from website

website_name = pagesoup.find('a',class_ = "navbar-brand")
link = "https://ncov2019.live/"
Markdown('<strong>{}</strong>{}'.format(website_name.text,link))

<strong>





    nCoV2019.live
  </strong>https://ncov2019.live/

In [15]:
#some quick facts from the website

quickfacts = pagesoup.find('div', class_ = "container--wrap bg-navy-4")
Markdown('<strong align="center">{}</strong>'.format(quickfacts))

<strong align="center"><div class="container--wrap bg-navy-4">
<p style="color: #DFDFEF; font-size: 30px; text-align: center; padding-top: 15px; font-weight: bolder; margin-bottom: 0;">
    Quick Facts
  </p>
<p style="color: #DFDFEF; font-size: 12px; text-align: center; font-weight: bolder; margin-bottom: 0;">
    updated: <i class="last_updated">A few minutes ago</i>
</p>
<table>
<tr>
<p style="color: rgb(101, 221, 155); font-weight: bold; text-align: center; border-bottom: 0; font-size: 50px; margin-bottom: 0px;">
        23,066,065
      </p>
<p style="color: #DFDFEF; font-size: 18px; font-weight: bold; border-bottom: 0; text-align: center;">
        Total Confirmed
      </p>
</tr>
<tr>
<p style="color: #ff9d00; font-weight: bold; text-align: center; border-bottom: 0; font-size: 50px; margin-bottom: 0px;"> 
          61,825
      </p>
<p style="color: #DFDFEF; font-size: 18px; font-weight: bold; border-bottom: 0; text-align: center;">
        Total Critical
      </p>
</tr>
<tr>
<p style="color: #F65164; font-weight: bold; text-align: center; border-bottom: 0; font-size: 50px; margin-bottom: 0px;">
        801,915
      </p>
<p style="color: #DFDFEF; font-size: 18px; font-weight: bold; border-bottom: 0; text-align: center;">
        Total Deceased
      </p>
</tr>
<tr>
<p style="color: rgb(248, 245, 64); font-weight: bold; text-align: center; border-bottom: 0; font-size: 50px; margin-bottom: 0px;">
        6,658,838
      </p>
<p style="color: #DFDFEF; font-size: 18px; font-weight: bold; border-bottom: 0; text-align: center;">
        Total Active
      </p>
</tr>
<tr>
<p style="color: rgb(68, 155, 226); font-weight: bold; text-align: center; border-bottom: 0; font-size: 50px; margin-bottom: 0px;">
        15,445,043
      </p>
<p style="color: #DFDFEF; font-size: 18px; font-weight: bold; border-bottom: 0; text-align: center;">
        Total Recovered
      </p>
</tr>
<tr>
<p style="color: rgb(62, 206, 229); font-weight: bold; text-align: center; border-bottom: 0; font-size: 50px; margin-bottom: 0px;">
        166<!--  / 43 -->
</p>
<p style="color: #DFDFEF; font-size: 18px; font-weight: bold; border-bottom: 0; text-align: center;">
        Total Vaccines In Development
      </p>
</tr>
</table>
</div></strong>

## World COVID-19 Stats

#### We will scrap worldwide covid cases.
1. We'll use pandas read.html which lets us read the webpage table without much of complexity.
2. Convert the table into dataframe for further processing.
3. In the header of the list generated you see a number "1", which was used in the original website as a filter for arranging data in ascending or descending order.

In [64]:
import pandas as pd
import requests

In [128]:
# grabbing latest worldwide data

url = "https://ncov2019.live/data/world"

r = requests.get(url)
df_list = pd.read_html(r.text)            #this parse all html tables from a webpage to alist
world_df = df_list[2]
world_df

Name  Confirmed Per Million  Changes Today  \
0                           TOTAL   23121366        2976          12950   
1                  ★  Afghanistan      37894         971              0   
2                      ★  Albania       8119        2822              0   
3                      ★  Algeria      40667         925              0   
4                      ★  Andorra       1045           0              0   
..                            ...        ...         ...            ...   
210      ★  Caribbean Netherlands         13           0              0   
211           ★  Falkland Islands         13           0              0   
212  ★  Saint Pierre and Miquelon          5           0              0   
213                   ★  Anguilla          3           0              0   
214                      ★  China      84939          59             22   

    Percentage Day Change Critical Deceased Per Million.1  Changes Today.1  \
0                   0.06%    61716   803223           103              623   
1                      0%       31     1385            35                0   
2                      0%       15      240            83                0   
3                      0%       41     1418            32                0   
4                      0%        1       53             0                0   
..                    ...      ...      ...           ...              ...   
210                    0%  Unknown  Unknown       Unknown                0   
211                    0%  Unknown  Unknown       Unknown                0   
212                    0%  Unknown  Unknown       Unknown                0   
213                    0%  Unknown  Unknown       Unknown                0   
214                 0.03%       20     4634             3                0   

    Percentage Death Change      Tests   Active Recovered Per Million.2  \
0                     0.08%  408073508  6662651  15509908          1996   
1                        0%     100377     8493     28016           718   
2                        0%      50036     3783      4096          1424   
3                        0%    Unknown  Unknown     28587           650   
4                        0%     137457      117       875             0   
..                      ...        ...      ...       ...           ...   
210                      0%        424  Unknown         7             0   
211                      0%       2256  Unknown        13             0   
212                      0%    Unknown  Unknown         1             0   
213                      0%        662  Unknown         3             0   
214                      0%   90410000      454     79851            55   

     Population  
0    7768876378  
1      39045465  
2       2877342  
3      43958573  
4         77283  
..          ...  
210       26257  
211        3493  
212        5790  
213       15022  
214  1439323776  

[215 rows x 15 columns]

In [129]:
# We will now sort the countries based on total confirmed cases column

world_df = world_df.sort_values("Confirmed" , ascending = False)



#Lets get top 10 affected countries

world_df.head(10)


Name  Confirmed Per Million  Changes Today  \
0               TOTAL   23121366        2976          12950   
169  ★  United States    5796727       17498              0   
170         ★  Brazil    3536488       16621              0   
171          ★  India    2975701        2153           2333   
172         ★  Russia     946976        6489              0   
142   ★  South Africa     603338       10155              0   
173           ★  Peru     576067       17438              0   
174         ★  Mexico     549734        4257           5928   
175       ★  Colombia     522138       10246              0   
177          ★  Chile     393769       20574              0   

    Percentage Day Change Critical Deceased Per Million.1  Changes Today.1  \
0                   0.06%    61716   803223           103              623   
169                    0%    16801   179200           541                0   
170                    0%     8318   113454           533                0   
171                 0.08%     8944    55928            40                0   
172                    0%     2300    16189           111                0   
142                    0%      539    12843           216                0   
173                    0%     1530    27245           825                0   
174                 1.09%     3518    59610           462              504   
175                    0%     1493    16568           325                0   
177                    0%     1072    10723           560                0   

    Percentage Death Change      Tests   Active Recovered Per Million.2  \
0                     0.08%  408073508  6662651  15509908          1996   
169                      0%   74718580  2490109   3127418          9440   
170                      0%   14144344   752279   2670755         12552   
171                      0%   34491073   697196   2222577          1608   
172                      0%   33800000   169457    761330          5217   
142                      0%    3504507    90393    500102          8417   
173                      0%    2915372   163914    384908         11651   
174                   0.85%    1239133   113715    376409          2915   
175                      0%    2409706   156630    348940          6848   
177                      0%    2142301    15149    367897         19222   

     Population  
0    7768876378  
169   331277530  
170   212772522  
171  1381900601  
172   145943480  
142    59412210  
173    33035148  
174   129121642  
175    50958476  
177    19139231

#### We'll use plotly express for visualization.
1. It generates graphs which are interactive and user friendly.
2. We can use zoom in and zoom out feature for proper understanding to a specific part of graph.

In [130]:
import plotly.express as px

In [131]:
# plotting world_df based on confirmed cases by country names.

world_fig = px.bar(world_df, x = 'Name' , y = 'Confirmed')
world_fig.show()

 - We can zoom in the graph, thats the beauty of plotly.

In [132]:
# Lets plot top 20 countries based on confirmed cases.

world_fig = px.bar(world_df.head(20), x = "Name" , y = 'Confirmed')
world_fig.show()

 - Now we can see United states holds number 1 position. (cough cough "we don't wear masks" - americans)
 - Brazil and India comes at the second and third position surpassing Russia respectively.

#### Now we'll try to explore the world_df in more details.

In [133]:
# Lets see how many people have died with respect to countries. (For top 20 countries)

world_fig = px.bar(world_df.head(20), x = 'Name', y = 'Confirmed', color = "Deceased")
world_fig.show()

 - Here the color of each bar corrosponds to how many people have died.
 - We cannot make out which country has most number of deceased people in a descending order.

In [134]:
# lets visualize world_df based on death poll for top 20 countries.

world_fig = px.bar(world_df.sort_values('Deceased'), x = 'Name', y = 'Deceased')
world_fig.show()

 -  As you see here also we cannot make countres based on number of deceased people.
 - There is some some problem with world_df["Deceased"] column, let's fix that.

In [135]:
# lets grab the world_df based on deceased column.
world_df.sort_values('Deceased',ascending = False)

Name  Confirmed Per Million  Changes Today  \
213        ★  Anguilla          3           0              0   
153     ★  Timor-Leste         26          20              0   
194   ★  Faroe Islands        384           0              0   
52          ★  Eritrea        306          86              0   
106        ★  Mongolia        298          91              0   
..                 ...        ...         ...            ...   
209      ★  Montserrat         13           0              0   
93    ★  Liechtenstein         99           0              0   
165  ★  Western Sahara         10           0              0   
201  ★  Cayman Islands        205           0              0   
56             ★  Fiji         28           0              0   

    Percentage Day Change Critical Deceased Per Million.1  Changes Today.1  \
213                    0%  Unknown  Unknown       Unknown                0   
153                    0%  Unknown  Unknown       Unknown                0   
194                    0%        4  Unknown       Unknown                0   
52                     0%  Unknown  Unknown       Unknown                0   
106                    0%        1  Unknown       Unknown                0   
..                    ...      ...      ...           ...              ...   
209                    0%  Unknown        1             0                0   
93                     0%  Unknown        1             0                0   
165                    0%  Unknown        1             0                0   
201                    0%  Unknown        1             0                0   
56                     0%  Unknown        1             1                0   

    Percentage Death Change    Tests   Active Recovered Per Million.2  \
213                      0%      662  Unknown         3             0   
153                      0%     5146  Unknown        25            19   
194                      0%    81991  Unknown       300             0   
52                       0%  Unknown  Unknown       274            77   
106                      0%    47587  Unknown       288            88   
..                      ...      ...      ...       ...           ...   
209                      0%       61        0        12             0   
93                       0%      900       10        88             0   
165                      0%  Unknown  Unknown         8             0   
201                      0%    34208        2       202             0   
56                       0%     6693        7        20             0   

    Population  
213      15022  
153    1321843  
194      48889  
52     3553187  
106    3285510  
..         ...  
209       4992  
93       38144  
165     599286  
201      65829  
56      897361  

[215 rows x 15 columns]

 - The column contains many unknown values.
 - We'll replace all the unknown values with zero.
 - Then we will arrange the column in descending order for visualization purpose.

In [136]:
# lets replace unknown values to 0.

world_df['Deceased'].replace("Unknown", 0,inplace=True)
world_df['Deceased'] = pd.to_numeric(world_df['Deceased'])          #convert column from type object to int64
world_df['Deceased']

0      803223
169    179200
170    113454
171     55928
172     16189
        ...  
211         0
70          0
165         1
212         0
213         0
Name: Deceased, Length: 215, dtype: int64

In [137]:
# now again lets grab world_df based on deceased column.

world_df.sort_values('Deceased',ascending = False)

Name  Confirmed Per Million  Changes Today  \
0                  TOTAL   23121366        2976          12950   
169     ★  United States    5796727       17498              0   
170            ★  Brazil    3536488       16621              0   
174            ★  Mexico     549734        4257           5928   
171             ★  India    2975701        2153           2333   
..                   ...        ...         ...            ...   
199  ★  French Polynesia        298           0              0   
106          ★  Mongolia        298          91              0   
52            ★  Eritrea        306          86              0   
194     ★  Faroe Islands        384           0              0   
213          ★  Anguilla          3           0              0   

    Percentage Day Change Critical  Deceased Per Million.1  Changes Today.1  \
0                   0.06%    61716    803223           103              623   
169                    0%    16801    179200           541                0   
170                    0%     8318    113454           533                0   
174                 1.09%     3518     59610           462              504   
171                 0.08%     8944     55928            40                0   
..                    ...      ...       ...           ...              ...   
199                    0%        2         0       Unknown                0   
106                    0%        1         0       Unknown                0   
52                     0%  Unknown         0       Unknown                0   
194                    0%        4         0       Unknown                0   
213                    0%  Unknown         0       Unknown                0   

    Percentage Death Change      Tests   Active Recovered Per Million.2  \
0                     0.08%  408073508  6662651  15509908          1996   
169                      0%   74718580  2490109   3127418          9440   
170                      0%   14144344   752279   2670755         12552   
174                   0.85%    1239133   113715    376409          2915   
171                      0%   34491073   697196   2222577          1608   
..                      ...        ...      ...       ...           ...   
199                      0%       9912  Unknown       148             0   
106                      0%      47587  Unknown       288            88   
52                       0%    Unknown  Unknown       274            77   
194                      0%      81991  Unknown       300             0   
213                      0%        662  Unknown         3             0   

     Population  
0    7768876378  
169   331277530  
170   212772522  
174   129121642  
171  1381900601  
..          ...  
199      281138  
106     3285510  
52      3553187  
194       48889  
213       15022  

[215 rows x 15 columns]

 - Perfecto!. Now we can see that column has been cleared off all the "Unknown".

In [138]:
# lets again try ro visualize world_df based on death poll for top 20 countries.

world_fig = px.bar(world_df.sort_values('Deceased', ascending=False).head(20), x = 'Name' , y = 'Deceased')
world_fig.show()


#### Click on the link for more information.

 - United States tops the chart. <a href = "https://www.sciencemag.org/news/2020/04/united-states-leads-coronavirus-cases-not-pandemic-response" > If you want to know why United States leads in coronvirus cases, but not pandemic response</a>
 - Brazil also surpasses 100,000 deaths and becomes the one of the worst affected countries. <a href = "https://www.ctvnews.ca/health/coronavirus/death-became-normal-brazil-surpasses-100-000-deaths-from-covid-19-1.5056757" >'Death became normal': Brazil surpasses 100,000 deaths from COVID-19</a>
 - Mexico's death toll also reached 59.106k and many young people are dying of COVID-19 <a href = "https://www.forbes.com/sites/nathanielparishflannery/2020/07/24/why-are-so-many-young-people-dying-of-covid-19-in-mexico-city/#148bc1f22792">Why Are So Many Young People Dying Of Covid-19 In Mexico City?</a>
 - India has also reached 56k and there are many questions about India's rising COVID-19 infection <a href ="https://www.bbc.com/news/world-asia-india-53018351">Five key questions about India's rising Covid-19 infections</a>

#### Lets visualize the death toll in relation to total confirmed case

In [139]:
# lets visualize the death toll based on total confirmed case

import plotly.graph_objects as go


# for grouped barplot using Deceased numbers per country and total number of cases per country.

fig = go.Figure(data = [
go.Bar(
    x = world_df['Name'],
    y = world_df["Deceased"].head(20),
    name = "Deceased",
    marker_color = "indianred"
),
go.Bar(
    x = world_df['Name'],
    y = world_df['Confirmed'].head(20),
    name = 'Confirmed',
    marker_color = "lightsalmon"
)
])

# Here we modify the tickangle of the xaxis, resulting in rotated labels.
fig.update_layout(barmode='group')
fig.show()

 - Here we can see the Death toll is very low as compared to confirmed cases, which is because most of the people recover from COVID-19. Early estimates predicted that the overall COVID-19 recovery rate is between 97% and 99.75%.
 - Mortality rate calculated  = 3.4% (802.318k/23.09665M)

#### lets visualize the recovered cases based on total confirmed case

In [140]:
# lets visualize the recovered case based in relation to total confirmed case

import plotly.graph_objects as go


# for grouped barplot using recovered cases per country and total number of cases per country.

fig = go.Figure(data = [
go.Bar(
    x = world_df['Name'],
    y = world_df["Recovered"].head(20),
    name = "Recovered",
    marker_color = "indianred"
),
go.Bar(
    x = world_df['Name'],
    y = world_df['Confirmed'].head(20),
    name = 'Confirmed',
    marker_color = "lightsalmon"
)
])

# Here we modify the tickangle of the xaxis, resulting in rotated labels.
fig.update_layout(barmode='group')
fig.show()

 - Here we can see how many person recovered in relation to total cases registered.
 - Recovery rate  = 67% (15.4827M/23.09665M), this contradicts early predicted value of recovery rate which was 97%.
 - Recovery rate and mortality rate are based on how well a country is implementing the testing of its people. <a href = "https://www.who.int/news-room/commentaries/detail/estimating-mortality-from-covid-19">Estimating mortality from COVID-19</a>

#### Lets see who has implemented testing vastly.

In [144]:
# replace unknown values from the column

world_df['Tests'].replace("Unknown", 0, inplace=True)
world_df['Tests'] = pd.to_numeric(world_df['Tests'])          #convert column from type object to int64



#Now lets plot the data

world_fig = px.bar(world_df.sort_values('Tests', ascending=False).head(20), x = 'Name' , y = 'Tests')
world_fig.show()

 - China on first position that was unexpected. I was expecting United States.
 - As you can see the countries who are vastly testing their people have a upper hand on curbing the spread of virus by implementing policies.

#### lets explore the Confirmed cases in relation to total population

In [119]:
# lets visualize the confirmed case based in relation to total population

import plotly.graph_objects as go


# for grouped barplot using confirmed cases per country and population per country.

fig = go.Figure(data = [
go.Bar(
    x = world_df['Name'],
    y = world_df["Confirmed"].head(20),
    name = "Confirmed",
    marker_color = "indianred"
),
go.Bar(
    x = world_df['Name'],
    y = world_df['Population'].head(20),
    name = 'Population',
    marker_color = "lightsalmon"
)
])

# Here we modify the tickangle of the xaxis, resulting in rotated labels.
fig.update_layout(barmode='group')
fig.show()

 - This graph shows a small percentage of people are affected by the novel coronavirus. <a href = "https://www.canada.ca/en/public-health/services/publications/diseases-conditions/people-high-risk-for-severe-illness-covid-19.html">
    People who are at high risk for severe illness from COVID-19</a>

## Canada COVID-19 Stats

#### Lets get Latest Canada's information.
1. We'll use the pandas read.html which lets us read the webpage table without much of complexity.
2. We can also use the lsit to convert it to a dataframe.
3. In the header of the list generated you see a number "1", which was used in the original website as a filter for arranging data in ascending or descending order.

In [62]:
#grabbing latest canada specific data

url = "https://ncov2019.live/data/canada"

r = requests.get(url)
df_list = pd.read_html(r.text) # this parses all the tables in webpages to a list
canada_df = df_list[2]
canada_df

Name  Confirmed Per Million  Changes Today  \
0                          TOTAL     125612     Unknown              0   
1                     ★  Alberta      12501     Unknown              0   
2            ★  British Columbia       4745     Unknown              0   
3                    ★  Manitoba        796     Unknown              0   
4               ★  New Brunswick        186     Unknown              0   
5   ★  Newfoundland and Labrador        268     Unknown              0   
6       ★  Northwest Territories          5     Unknown              0   
7                 ★  Nova Scotia       1077     Unknown              0   
8                     ★  Ontario      42983     Unknown              0   
9        ★  Prince Edward Island         44     Unknown              0   
10                     ★  Quebec      61402     Unknown              0   
11               ★  Saskatchewan       1590     Unknown              0   
12                      ★  Yukon         15     Unknown              0   

   Percentage Day Change Critical Deceased Per Million.1  Changes Today.1  \
0                     0%       61     9096       Unknown                0   
1                     0%  Unknown      227       Unknown                0   
2                     0%  Unknown      198       Unknown                0   
3                     0%  Unknown       12       Unknown                0   
4                     0%  Unknown        2       Unknown                0   
5                     0%  Unknown        3       Unknown                0   
6                     0%  Unknown  Unknown       Unknown                0   
7                     0%  Unknown       64       Unknown                0   
8                     0%  Unknown     2838       Unknown                0   
9                     0%  Unknown  Unknown       Unknown                0   
10                    0%  Unknown     5730       Unknown                0   
11                    0%  Unknown       22       Unknown                0   
12                    0%  Unknown  Unknown       Unknown                0   

   Percentage Death Change    Tests   Active  Recovered Per Million.2  \
0                       0%  4974215     4717     111798       Unknown   
1                       0%  Unknown  Unknown      11167       Unknown   
2                       0%  Unknown  Unknown       3749       Unknown   
3                       0%  Unknown  Unknown        537       Unknown   
4                       0%  Unknown  Unknown        178       Unknown   
5                       0%  Unknown  Unknown        263       Unknown   
6                       0%  Unknown  Unknown          5       Unknown   
7                       0%  Unknown  Unknown       1007       Unknown   
8                       0%  Unknown  Unknown      39035       Unknown   
9                       0%  Unknown  Unknown         40       Unknown   
10                      0%  Unknown  Unknown      54383       Unknown   
11                      0%  Unknown  Unknown       1419       Unknown   
12                      0%  Unknown  Unknown         15       Unknown   

   Population  
0     Unknown  
1     Unknown  
2     Unknown  
3     Unknown  
4     Unknown  
5     Unknown  
6     Unknown  
7     Unknown  
8     Unknown  
9     Unknown  
10    Unknown  
11    Unknown  
12    Unknown

## Canada COVID-19 Stats

#### Lets visualize Canada's Data and see which province has been worst effected.
1. We'll use the same above canada_df for visualization purpose.
2. We are going to use this dataframe because it's the latest data and our script we'll update the data every time we run the cell based on the website mentioned above.
3. I'm going to use plotly for visualization purpose as it generates graphs which are interactive and user friendly.

In [73]:
canada_fig = px.bar(canada_df.sort_values('Confirmed'),x='Name',y='Confirmed',color="Deceased")
canada_fig.show()

 - Overall canada has 125.612k confirmed cases with a death toll of 9096.
 - As we can see Quebec has been first affected with death toll of 5730.
 - Northwest Territories has not been affected much with only 5 confirmed cases.
 - Alberta has 12.50k confirmed cases with a death toll of 227.